core/amber/src/main/python/core/models/marker.py

```diff
@@ -1,11 +1,70 @@
 from dataclasses import dataclass
+from pyarrow import Table
+from pandas import DataFrame
+from typing import Optional
+from .schema import Schema, AttributeType
+from .schema.attribute_type import FROM_PYOBJECT_MAPPING
 
 
 @dataclass
 class Marker:
     pass
 
 
+@dataclass
+class StartOfUpstream(Marker):
+    pass
+
+
 @dataclass
 class EndOfUpstream(Marker):
     pass
+
+
+@dataclass
+class State(Marker):
+    def __init__(
+        self, table: Optional[Table] = None, pass_to_all_downstream: bool = False
+    ):
+        if table is None:
+            self.data = {}
+            self.schema = Schema()
+            self.add("passToAllDownstream", pass_to_all_downstream)
+        else:
+            self.data = table.to_pandas().iloc[0].to_dict()
+            self.schema = Schema(table.schema)
+
+    def add(
+        self, key: str, value: any, value_type: Optional[AttributeType] = None
+    ) -> None:
+        self.data[key] = value
+        if value_type is not None:
+            self.schema.add(key, value_type)
+        else:
+            self.schema.add(key, FROM_PYOBJECT_MAPPING[type(value)])
+
+    def get(self, key: str) -> any:
+        return self.data[key]
+
+    def is_pass_to_all_downstream(self) -> bool:
+        return self.data["passToAllDownstream"]
+
+    def to_table(self) -> Table:
+        return Table.from_pandas(
+            df=DataFrame([self.data]),
+            schema=self.schema.as_arrow_schema(),
+        )
+
+    def __setitem__(self, key: str, value: any, value_type: AttributeType) -> None:
+        self.add(key, value, value_type)
+
+    def __getitem__(self, key: str) -> any:
+        return self.get(key)
```

Comment:
just store in `self.__dict__` and no need to overwrite those functions.

Reply:
Moved data to self.__dict__, but we still want to override those functions such that the user can access a state by using state["something"].